From edffc1f8d02966989aa5b0c2d13b2f4d0c96e8aa Mon Sep 17 00:00:00 2001
From: Chenhui Zhao <chenhui.zhao@freescale.com>
Date: Wed, 23 Apr 2014 16:24:14 +0800
Subject: [PATCH 1162/1207] powerpc/deepsleep: Disable CPC speculation to avoid
 deep sleep hang

[Original patch taken from QorIQ-SDK-V1.6-SOURCE-20140619-yocto.iso]

Disable CPC speculation to avoid deep sleep hang, especially
in secure boot mode. This bit will be cleared automatically
when resuming from deep sleep.

Change-Id: I2e4debed63eed39fa059c42de5b7f49b5d6c48f3
Signed-off-by: Chenhui Zhao <chenhui.zhao@freescale.com>
Reviewed-on: http://git.am.freescale.net:8181/11362
Tested-by: Review Code-CDREVIEW <CDREVIEW@freescale.com>
Reviewed-by: Dongsheng Wang <dongsheng.wang@freescale.com>
Reviewed-by: Yang Li <LeoLi@freescale.com>
Reviewed-by: Jose Rivera <German.Rivera@freescale.com>
---
 arch/powerpc/platforms/85xx/deepsleep.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/arch/powerpc/platforms/85xx/deepsleep.c b/arch/powerpc/platforms/85xx/deepsleep.c
index 4a41020..88dee90 100644
--- a/arch/powerpc/platforms/85xx/deepsleep.c
+++ b/arch/powerpc/platforms/85xx/deepsleep.c
@@ -21,6 +21,9 @@
 #define SIZE_1MB	0x100000
 #define SIZE_2MB	0x200000
 
+#define CPC_CPCHDBCR0		0x10f00
+#define CPC_CPCHDBCR0_SPEC_DIS	0x08000000
+
 #define CCSR_SCFG_DPSLPCR	0xfc000
 #define CCSR_SCFG_DPSLPCR_WDRR_EN	0x1
 #define CCSR_SCFG_SPARECR2	0xfc504
@@ -169,6 +172,13 @@ int fsl_enter_epu_deepsleep(void)
 	clrbits32(ccsr_base + CCSR_GPIO1_GPODR, CCSR_GPIO1_GPDIR_29);
 	setbits32(ccsr_base + CCSR_GPIO1_GPDIR, CCSR_GPIO1_GPDIR_29);
 
+	/*
+	 * Disable CPC speculation to avoid deep sleep hang, especially
+	 * in secure boot mode. This bit will be cleared automatically
+	 * when resuming from deep sleep.
+	 */
+	setbits32(ccsr_base + CPC_CPCHDBCR0, CPC_CPCHDBCR0_SPEC_DIS);
+
 	fsl_dp_fsm_setup(dcsr_base, NULL);
 
 	fsl_dp_enter_low(ccsr_base, dcsr_base, pld_base, pld_flag);
-- 
2.0.2

